graylog2-server/src/main/java/org/graylog2/system/shutdown/GracefulShutdownService.java

```diff
@@ -0,0 +1,139 @@
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A service that participates in the Graylog server graceful shutdown.
+ * <p>
+ * Services can implement {@link GracefulShutdownHook} and register themselves with this service to make sure they
+ * get shut down properly on server shutdown. During shutdown the registered hooks will be called in no particular
+ * order.
+ * <p>
+ * Make sure to use {@link #unregister(GracefulShutdownHook)} if a registered service is shutting down before the
+ * server shutdown to avoid leaking service instances in the {@link GracefulShutdownService}.
+ *
+ * See {@link GracefulShutdownHook} for an example.
+ */
+@Singleton
+public class GracefulShutdownService extends AbstractIdleService {
+    private static final Logger LOG = LoggerFactory.getLogger(GracefulShutdownService.class);
+
+    private final Set<GracefulShutdownHook> shutdownHooks = ConcurrentHashMap.newKeySet();
+    private final AtomicBoolean isShuttingDown = new AtomicBoolean(false);
+
+    @Override
+    protected void startUp() {
+        // Nothing to do
+    }
+
+    @Override
+    protected void shutDown() {
+        // Don't do anything if the shutdown is already in progress or if there are no hooks registered
+        if (isShuttingDown.getAndSet(true) || shutdownHooks.isEmpty()) {
+            return;
+        }
+
+        try {
+            // Use an executor to run the shutdown hooks in parallel but don't start too many threads
+            // TODO: Make max number of threads user configurable
+            final ExecutorService executor = executorService(Math.min(shutdownHooks.size(), 10));
+            final CountDownLatch latch = new CountDownLatch(shutdownHooks.size());
+
+            LOG.info("Running graceful shutdown for <{}> shutdown hooks", shutdownHooks.size());
+            for (final GracefulShutdownHook shutdownHook : shutdownHooks) {
+                executor.submit(() -> {
+                    final String hookName = shutdownHook.getClass().getSimpleName();
+                    try {
+                        LOG.info("Initiate shutdown for <{}>", hookName);
+                        final Stopwatch stopwatch = Stopwatch.createStarted();
+                        shutdownHook.doGracefulShutdown();
+                        LOG.info("Finished shutdown for <{}>, took {} ms", hookName, stopwatch.stop().elapsed(TimeUnit.MILLISECONDS));
+                    } catch (Exception e) {
+                        LOG.error("Problem shutting down <{}>", hookName, e);
+                    } finally {
+                        latch.countDown();
+                    }
+                });
+            }
+
+            latch.await();
+            executor.shutdown();
+            executor.awaitTermination(1, TimeUnit.MINUTES);
```

Comment:
I don't think is necessary here, if the latch is open, all tasks have completed and we don't need to wait for anything.

Reply:
Agreed, thanks! :+1: